mycore-base/src/main/java/org/mycore/frontend/jersey/filter/MCRSessionHookFilter.java

```diff
@@ -47,20 +48,68 @@
     @Context
     private HttpServletResponse httpResponse;
 
+    private final static String ATTR = MCRSessionHookFilter.class.getName() + ".session";
+
     private static final Logger LOGGER = LogManager.getLogger(MCRSessionHookFilter.class);
 
     @Override
-    public void filter(ContainerRequestContext request) throws IOException {
+    public void filter(ContainerRequestContext request) {
+        MCRSessionMgr.unlock();
         MCRSession session = MCRServlet.getSession(httpRequest);
+        request.setProperty(ATTR, session);
         MCRSessionMgr.setCurrentSession(session);
         LOGGER.info(MessageFormat.format("{0} ip={1} mcr={2} user={3}", request.getUriInfo().getPath(),
             MCRFrontendUtil.getRemoteAddr(httpRequest), session.getID(), session.getUserInformation().getUserID()));
         MCRFrontendUtil.configureSession(session, httpRequest, httpResponse);
     }
 
     @Override
-    public void filter(ContainerRequestContext request, ContainerResponseContext response) throws IOException {
-        MCRSessionMgr.releaseCurrentSession();
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
+        MCRSessionMgr.unlock();
+        MCRSession requestSession = (MCRSession) requestContext.getProperty(ATTR);
+        if (responseContext.hasEntity()) {
+            responseContext.setEntityStream(new ProxyOutputStream(responseContext.getEntityStream()) {
+                @Override
+                public void close() throws IOException {
+                    LOGGER.debug("Closing EntityStream");
+                    try {
+                        super.close();
+                    } finally {
+                        releaseSessionIfNeeded(requestSession);
+                        LOGGER.debug("Closing EntityStream done");
+                    }
+                }
+            });
+        } else {
+            LOGGER.debug("No Entity in response, closing MCRSession");
+            releaseSessionIfNeeded(requestSession);
+        }
+    }
+
+    private static void releaseSessionIfNeeded(MCRSession requestSession) {
+        if (MCRSessionMgr.hasCurrentSession()) {
+            MCRSession currentSession = MCRSessionMgr.getCurrentSession();
+            try {
+                if (currentSession.isTransactionActive()) {
+                    LOGGER.debug("Active MCRSession and JPA-Transaction found. Clearing up");
+                    if (currentSession.transactionRequiresRollback()) {
+                        currentSession.rollbackTransaction();
+                    } else {
+                        currentSession.commitTransaction();
+                    }
+                } else {
+                    LOGGER.debug("Active MCRSession found. Clearing up");
+                }
+            } finally {
+                MCRSessionMgr.releaseCurrentSession();
+                if (!currentSession.equals(requestSession)) {
```

Comment:
why can a jersey session be not bound to an HTTP session? I seems that we fix here symptoms and not the root cause if it is so.

Reply:
> why can a jersey session be not bound to an HTTP session? I seems that we fix here symptoms and not the root cause if it is so.

There are 3 phases and this class handles 2.

1. An MCRSession (A) is optained or created and bound to HttpSession and stored in the current request. When the HttpSession hits EOL the MCRSession is closed, too.
1. Other request or response filters take over and the resource code is executed.
1. The response is filtered and we check if session A is still used or was switched in phase 2. If that is the case the session will leak as it is not bound to the HttpSession. We close it for that reason.